File: Python/Programmers/네트워크.py
from collections import deque

def solution(n, computers):
    answer = 0

    visited = [0] * n # 방문을 확인하기 위해 크기가 n 리스트 선언
    queue = deque()
    while 0 in visited: # 모든 컴퓨터가 연결되어 있지 않으므로 모든 컴퓨터가 방문 되도록 반복
        queue.append(visited.index(0)) # 처음 0의 위치
        while queue:
            v = queue.popleft()
            for i in range(n):
                if visited[i] == 0 and computers[v][i] == 1: # 아직 방문하지 않았고 컴퓨터 값이 1이라면
                    queue.append(i)
                    visited[i] = 1
        answer += 1 # 연결된 컴퓨터가 없으면 +1

    return answer

File: JavaScript/Programmers/Level 2/네트워크.js
// dfs
// function solution(n, computers) {
//     const visited = Array.from({ length: n }, () => false)
    
//     const dfs = (i) => {
//         visited[i] = true
//         for (let j = 0; j < computers[i].length; j++) {
//             if (!visited[j] && computers[i][j] === 1) {
//                 dfs(j)
//             }
//         }
//     }
    
//     let answer = 0
//     for (let i = 0; i < n; i++) {
//         if (!visited[i]) {
//             answer += 1
//             dfs(i)
//         }
//     }
    
//     return answer
// }

// bfs
function solution(n, computers) {
    const visited = Array.from({ length: n }, () => false)
    
    const bfs = (i) => {
        visited[i] = true
        const queue = [i]
        while (queue.length > 0) {
            const v = queue.shift()
            for (let j = 0; j < computers[v].length; j++) {
                if (!visited[j] && computers[v][j] === 1) {
                    visited[j] = true
                    queue.push(j)
                }
            }
        }
    }
    
    let answer = 0
    for (let i = 0; i < computers.length; i++) {
        if (!visited[i]) {
            answer += 1
            bfs(i)
        }
    }
    
    return answer
}